lib/core/src/Cardano/Wallet/Api/Types.hs

```diff
@@ -997,11 +1016,42 @@ instance DecodeAddress n => FromJSON (ApiSelectCoinsData n) where
 instance EncodeAddress n => ToJSON (ApiSelectCoinsData n) where
     toJSON = genericToJSON defaultRecordTypeOptions
 
+instance DecodeAddress n => FromJSON (ApiSelectCoinsActionData n) where
+    parseJSON = withObject "ApiSelectCoinsActionData" $ \obj ->
+        (joinAction obj <|> quitAction obj)
+      where
+        joinAction obj = do
+          pass <- obj .: "passphrase"
+          ("join" :: String) <- obj .: "action"
+          pid <- obj .: "pool"
+          pure $ ApiSelectCoinsActionData (Join (getApiT pid)) pass
+        quitAction obj = do
+          pass <- obj .: "passphrase"
+          ("quit" :: String) <- obj .: "action"
+          pure $ ApiSelectCoinsActionData Quit pass
+
+instance EncodeAddress n => ToJSON (ApiSelectCoinsActionData n) where
+    toJSON (ApiSelectCoinsActionData Quit pass)
+        = object [ "action" .= String "quit", "passphrase" .= pass ]
+    toJSON (ApiSelectCoinsActionData (Join poolid) pass)
+        = object [ "action" .= String "join", "pool" .= ApiT poolid, "passphrase" .= pass ]
```

Comment:
These would be better derived using Generics. Hand-written JSON usually leads to inconsistencies and are hard to maintain on large APIs.

Reply:
The representation is too different from the data. That's what json roundtrip tests are for.